src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.cs

```diff
@@ -980,9 +980,14 @@ private void InvalidateTextBlock()
 
 		public override string GetAccessibilityInnerText() => Text;
 
-		// This approximates UWP behavior
-		private protected override double GetActualWidth() => DesiredSize.Width;
-		private protected override double GetActualHeight() => DesiredSize.Height;
+		// On UWP, ActualSize is immediately available after the TextBlock is measured, even if the TextBlock
+		// is never added to the visual tree. In Uno, we just return a best effort size until the TextBlock is
+		// attached to the visual tree and then we can depend on the layout cycle to return the proper ActualSize
+		private protected override double GetActualWidth()
+			=> !IsFirstArrangeDone ? (DesiredSize.Width - Margin.Left - Margin.Right).AtLeast(MinWidth).AtMost(MaxWidth) : base.GetActualWidth();
```

Comment:
Does this include an leave/enter cycle (Where a control loses its parent), or maybe a load/unload? We should probably add tests for those scenario as well.

Reply:
We currently have a problem with unloading and DesiredSize. As we discussed, I opened #13299 instead of tackling it here.